### PR TITLE
CA2263: Prefer generic overload when type is known

### PR DESCRIPTION
**Pull request by mpidash**
Fixes dotnet/runtime#68243.

Category: Usage
Severity: Info

This analyzer detects when a `System.Type` overload is called when a suitable generic overload is available and the type is known at compile-time. The fixer replaces it with the generic call and removes redundant casts and parentheses.

To validate if a generic overload is applicable, the following rules are checked:
1. The arity of the generic candidate must match the number of type arguments
2. The number of parameters must match the number of the other arguments (arguments minus type arguments) of the original call.
3. The candidate must not be the same as the enclosing symbol of the original call (see below for why)
4. The return types must be compatible (i.e., the return type of the candidate method must be assignable to the return type of the original call). For expression statements, the return type check is skipped.
5. The other arguments must be compatible (i.e. the other arguments of the original call must be assignable to the parameters of the generic candidate).
6. A speculative binding of the modified call must succeed. This is to check if any generic type constraints are violated.

---

Some notes and corner cases found during testing:

Rule 3. is due to the following [case](https://github.com/dotnet/runtime/blob/79c021d65c280020246d1035b0e87ae36f2d36a9/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L182) used in the runtime repository:

```C#
public static IntPtr OffsetOf<T>(string fieldName) => OffsetOf(typeof(T), fieldName);
```

If we did not check the enclosing symbol, the fixer would generate an endless recursion.

Rule 4. is due to the following [case](https://github.com/dotnet/roslyn/blob/71c2b3d8977d0cd865133993694d9c3c07ae9608/src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/NameSyntaxClassifier.cs#L36):

```C#
public override ImmutableArray<Type> SyntaxNodeTypes { get; } = ImmutableArray.Create(typeof(NameSyntax));
```

Here, the intention is to create a `ImmutableArray<Type>` with one element `typeof(NameSyntax)`. If the analyzer would not check the return value, this would lead to a diagnostic and would be fixed to `ImmutableArray.Create<NameSyntax>()` which returns an empty `ImmutableArray<NameSyntax>`.

Rule 6. prevents compile errors due to type constraints, for [example](https://github.com/dotnet/roslyn/blob/71c2b3d8977d0cd865133993694d9c3c07ae9608/src/Compilers/Core/Portable/InternalUtilities/EnumUtilties.cs#L53-L56):

```C#
internal static T[] GetValues<T>() where T : struct
{
    return (T[])Enum.GetValues(typeof(T));
}
```

We cannot change to `return Enum.GetValues<T>()` as there is only a struct type constraint, but the generic version of `Enum.GetValues` also constrains `T` to the `Enum` type (in addition to `struct`).

The tests show all other corner cases that I've encountered during testing.

---

I have one open question: How should we handle different runtime targets? For example the following

```C#
internal static T[] Foo<T>() where T : struct, Enum
{
    return (T[])Enum.GetValues(typeof(T));
}
```

would produce a diagnostic when targeting `net7.0`, but not when targeting `net472`, as Enum.GetValues<T> is only available since `net5`. Is this ok because the default severity is info?

---

I've also added all findings from `roslyn-analyzer`, `roslyn` and `runtime` (set severity to `none` for tests in `runtime`, rule ID is still CA2262 in the files):
[roslyn-analyzer.txt](https://github.com/dotnet/roslyn-analyzers/files/12353441/roslyn-analyzer.txt)
[roslyn-cs.txt](https://github.com/dotnet/roslyn-analyzers/files/12353442/roslyn-cs.txt)
[roslyn-vb.txt](https://github.com/dotnet/roslyn-analyzers/files/12353439/roslyn-vb.txt)
[runtime-clr-mono-libs-libs.tests.txt](https://github.com/dotnet/roslyn-analyzers/files/12353440/runtime-clr-mono-libs-libs.tests.txt)

### CODE REVIEW

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpPreferGenericOverloads.Fixer.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    public sealed class CSharpPreferGenericOverloadsFixer : PreferGenericOverloadsFixer
+    {
+        protected override async Task<Document> ReplaceWithGenericCallAsync(Document document, IInvocationOperation invocation, CancellationToken cancellationToken)
+        {
+            if (!RuntimeTypeInvocationContext.TryGetContext(invocation, out var invocationContext))
+            {
+                return document;
+            }
```
Comment (Youssef1313) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpPreferGenericOverloads.Fixer.cs:30]: Is this condition expected to be hit? I assume it shouldn't because the analyzer should have already filtered valid invocations only?

Comment (mpidash) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpPreferGenericOverloads.Fixer.cs:30]: I would not expect it to be hit, but I was not sure if there is a corner case, where the symbols are found in the analyzer, but are not found when the fixer is run. Also, this silences the nullability warnings when I use the context later.

Should I remove the check or add a comment that this branch is very unlikely to be hit?

Comment (Youssef1313) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Usage/CSharpPreferGenericOverloads.Fixer.cs:30]: The code is probaby okay as is. I was just making sure this is not expected to be hit.

**Comment by buyaa-n** [comment]
> Rule 3. is due to the following [case](https://github.com/dotnet/runtime/blob/79c021d65c280020246d1035b0e87ae36f2d36a9/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs#L182) used in the runtime repository:

Good catch 👍 

> 4. The return types must be compatible (i.e., the return type of the candidate method must be assignable to the return type of the original call). For expression statements, the return type check is skipped.
> 5. The other arguments must be compatible (i.e. the other arguments of the original call must be assignable to the parameters of the generic candidate).

Shouldn't we restrict to exact match not compatible to avoid false positives? Is there a specific scenario that needs to be compatible?

> How should we handle different runtime targets?

By my understanding the analyzer compilation only have the current target info, could not give info for other targets. Now all similar analyzers cause this issue and in runtime we if-def the solution in case the analyzer is set for warning. I don't aware of any solution for this, @mavasani might not a solution

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/PreferGenericOverloadsTests.cs [hunk 2]
```diff
@@ -0,0 +1,2409 @@
...
+        }
+
+        [Fact]
+        public async Task SingleTypeArgumentWithOtherGenericArgument_OffersFixer_CS()
+        {
+            string source = """
+                class C
+                {
+                    void M(System.Type type, object x) {}
+                    void M<T>(T x) {}
```
Comment (buyaa-n) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/PreferGenericOverloadsTests.cs:698]: Wonder if this scenario cause more false positives ...

Comment (mpidash) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/PreferGenericOverloadsTests.cs:698]: IIRC this test was motivated by the following (from roslyn-cs.txt):
W:\roslyn\src\Compilers\CSharp\Test\Semantic\SourceGeneration\GeneratorDriverTests.cs(3102,78): info CA2262: Prefer the generic overload 'System.Enum.GetName<TEnum>(TEnum)' instead of 'System.Enum.GetName(System.Type, object)'

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Usage/PreferGenericOverloadsTests.cs:698]: `System.Enum.GetName<TEnum>(TEnum)` instead of `System.Enum.GetName(System.Type, object)` is a good example, thank you!

**Review by buyaa-n — APPROVED** [review]
The PR looks great, thank you @mpidash!

Test coverage looks good, many hits in runtime, but with a quick check did not see a false positive, the fixer works as expected, good to go
